### ISSUE TITLE
Adding Entropy instead of dT/dP

### ISSUE THREAD

**Issue by Michlozz** [Huang-CL/Magrathea#11]
The program runs with my new EOS, but there is one more issue: temperature profile.

Now it runs with  " dTdP_QEOS( P,  T)= 2.*T / (7.*P); " (a copy from dTdP_gas). This approximation is not very good for our case.

In a paper you write: "In addition, either a temperature gradient, dT/dP, as a function of pressure and temperature or an entropy function dependent on density and temperature can be used to set up the temperature solver in a phase."

So, if I understand correctly, I can provide S(T,rho) function **instead** of dTdP_QEOS, right? How do I do it?

**Comment by Huang-CL** [comment]
Yes, you can create a `S_Si_QEOS(rho, T)` function. And then call `Si_QEOS->modify_extern_entropy(S_Si_QEOS);`.

**Comment by Michlozz** [comment]
It seem to work, but I got some trouble in the function "EOS::dTdV_S"(EOS.cpp).
The line "return - pSpV_T(V,T) / pSpT_V(V,T);" gave nan, as both of the derivatives were zero.
In both of the functions the line
 "gsl_deriv_central(&F, V, 1E-4, &result, &abserr);" 
 gave zero result.
I replaced it by:
 "res=(S_T(V,&params)-S_T(V+dV,&params))/dV;"

**Comment by Huang-CL** [comment]
gsl_deriv_central has an integer return value (flag) that returns the error code.  You may check it with 
```
int status;
status = gsl_deriv_central(&F, V, 1E-4, &result, &abserr);
cout<<"Error code: "<<gsl_strerror (status)<<' '<<status<<endl;
```
It can also be helpful for you to debug the problem.

**Comment by Michlozz** [comment]
The return status was 0 (success), so the function works. I changed the step from 1E-4 to 1E0, and it changed the value of "result" form 0 to 224253, the same order of magnitude as I get form the "manual" derivative method.
And this leads me to a question, what **units** and order of magnitude you expect at ~600K and ~200Gpa?
I have S_T ~1e7 erg/(gram*K). Does it make sense?

**Comment by Huang-CL** [comment]
It's weird. A classic value for V is on the order of 10 cm^3/mol. Absolute stepsize of 1E-4 indicates a relative step size of 1E-5. I believe it is an appropriate value. I don't see why it fails.  Can you also print `abserr` with both of your set up?

Because the unit of entropy is always canceled in the calculation, the code does not require a specific unit.  However, the other part of the calculation uses the S/(nR) for entropy, where n is the number of atoms per cell, and R is the gas constant.  Therefore, the entropy used in the code is a dimensionless quantity.

I tested with the built-in Fo EOS at 600K and 200GPa.  I got rho=4.61 g/cm^3,  V=30.5155 cm^3/mol, S=1.314, 
pSpV_T = 0.03356 mol/cm^3, pSpT_V = 0.003554 K^-1, dTdV_S = -9.441 K mol/cm^3.

**Comment by Michlozz** [comment]
In **double EOS::pSpV_T(double V, double T)** with:
T=644.957
V=13.943
I get
for:
_gsl_deriv_central(&F, V, 1e-4, &result, &abserr)_ I get
abserr=0.000611563
result=0

and for: 
_gsl_deriv_central(&F, V, 1e-1, &result, &abserr)_ I get
abserr=23392.9
result=567275

What I see immediately form your answer is different orders of magnitude. While you talk about S=1.3, while I got S_T(V,&params)=4.59039e+07 erg/(gr*K). These are the typical entropy values I get from EOS, and thus pSpV_T is 7 orders of magnitude higher. This may be the reason I get bad results.
How did you normalize the a units to be dimensionless quantity? I guess you normalize by some S0.

**Comment by Huang-CL** [comment]
Normalize by dividing nR, where n is the number of atoms per cell, and R is the gas constant.

**Comment by Huang-CL** [comment]
If I were you, I would add print statement in S_T to get V, T, rho, and entropy, and compare the outputs when you choose different initial stepsize.

**Comment by Michlozz** [comment]
I divided the S by (n=3)*(R), to be in the same orders of magnitude. I adjusted the step to be 0.5% of the input value, for example  "gsl_deriv_central(&F, V, dV, &result, &abserr);", where dV=0.005*V.

In EOS::dTdV_S, I have:
P=192.186
T=644.957
V=13.943
thermal_type=1
pSpV_T(V,T)=0.00221964
pSpT_V(V,T)=0.000875548

inside the derivatives:
In EOS::pSpV_,T I have:
abserr=1.22967e-10
result=0.00221964
S_T(V,&params)=0.184032

In EOS::pSpT_V, I have:
abserr=1.76133e-06
result=0.000875548
S_V(T,&params)=0.184032

Therefore, - pSpV_T(V,T) / pSpT_V(V,T)=-2.53515

The code runs, and creates output file, so it's a good sign. However, the output surprises me, as I get isothermal profile instead of adiabatic profile. 

One more thing that disturbs me it that my s function I built (called S_mix(rho,T) ) sometimes get input values much higher than shown in the output file. For example, T=644.95 and rho=7.97. These values pretty far from the final results (of T=541, rho=2.61-4.62). The function buit in away that for high values of rho (my table goes up to ~8 gr/cm^3), I set s to be constant.

Can these two problems be related?

[Structure27_7_2023.txt](https://github.com/Huang-CL/Magrathea/files/12185090/Structure27_7_2023.txt)

**Comment by Huang-CL** [comment]
`pPprho_T(rho,T)` and `pPpT_rho(rho,T)` are also necessary to get `dTdm` used in the temperature gradient equation. (see line 1406 in EOS.cpp)
Since you encountered problem in `pSpV_T` and `pSpT_V`, I suspect you may have similar problem with `pPprho_T` (line 1344) and `pPpT_rho` (line 1357 EOS.cpp).  Please check their output.

 The high density may occur when the equation of state have multiple density solutions, and the code is confused in which branch of the solution to pick.   I made some patches to reduce the occurrence rate by teaching the code how to select the correct branch, but it may still happen, esp. when the solution run outside the expected parameter envelope.  Therefore, I hope the high density problem will go away after you fix the temperature profile problem.

**Comment by Huang-CL** [comment]
David told me earlier that your entropy function is a high order polynomial.  If you can write the derivative function analytically, it is absolutely preferred over using the built-in numerical method.

**Comment by Michlozz** [comment]
I integrated a "  Si_QEOS->modify_dTdP(dTdP_mix);", there dTdP_mix is defined via "double dTdP_mix(double P, double T)" and it calls few functions to find polynomial coefficients
